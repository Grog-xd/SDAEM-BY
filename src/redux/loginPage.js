import {createSlice} from '@reduxjs/toolkit';

import avatar from '../assets/img/avatar.png'

const loginPage = createSlice({
    name: 'login',
    initialState:{
        // bookMarkActive:false,
        isAuth: false,
        profilesArr:[{name: 'dmitriy', avatar:'', phone:'', city:'', email: 'vladimir6234@tut.by', password: '12345', viber: '', whatsUpp:''},{name: 'test', avatar:avatar, phone:'+test', city:'test', email: 'test@tut.by', password: 'test1', viber: 'test', whatsUpp:'test'}],
        profile: {},
    },
    reducers:{
        // setBookMarkActive(state){
        //     state.bookMarkActive = !state.bookMarkActive
        // },
        enter(state, action){
            state.isAuth = true
            state.profile = action.payload
        },
        exit(state){
            state.isAuth = false
            localStorage.clear()
        },
        registerUser(state, action){
            state.profilesArr = [...state.profilesArr, action.payload]
        },
    },
})

export default loginPage.reducer
export const {exit, enter, setBookMarkActive, registerUser} = loginPage.actions

