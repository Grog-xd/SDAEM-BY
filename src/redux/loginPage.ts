import {createSlice, PayloadAction} from '@reduxjs/toolkit';

import avatar from '../assets/img/avatar.png'
import {profilesArr} from '../types/types';

const loginPage = createSlice({
    name: 'login',
    initialState:{
        // bookMarkActive:false,
        isAuth: false,
        profilesArr:[{name: 'dmitriy', avatar:'', phone:'', city:'', email: 'vladimir6234@tut.by', password: '12345', viber: '', whatsUpp:''},{name: 'test', avatar:avatar, phone:'+test', city:'test', email: 'test@tut.by', password: 'test1', viber: 'test', whatsUpp:'test'}],
        profile: {},
    },
    reducers:{
        // setBookMarkActive(state){
        //     state.bookMarkActive = !state.bookMarkActive
        // },
        enter(state, action:PayloadAction<profilesArr>){
            state.isAuth = true
            state.profile = action.payload
        },
        exit(state){
            state.isAuth = false
            localStorage.clear()
        },
        registerUser(state, action:PayloadAction<any>){
            state.profilesArr = [...state.profilesArr, action.payload]
        },
    },
})

export default loginPage.reducer
export const {exit, enter, registerUser} = loginPage.actions

