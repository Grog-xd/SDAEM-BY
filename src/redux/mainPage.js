import {createSlice} from '@reduxjs/toolkit';

const main = createSlice({
    name: 'main',
    initialState:{
        tabs:[{id:'flats', value:'Квартиры'}, {id:'cottages', value:'Котеджы и усадьбы'}, {id:'baths', value:'Бани и сауны'},  {id:'cars', value:'Авто на прокат'}],

        cityOption:[{id:1,value:'Минск'},{id:2,value:'Гомель'},{id:3,value:'Брест'},{id:4,value:'Витебск'},{id:5,value:'Гродно'}, {id:6,value:'Могилев'}],
        cityCurrentValue:'',

        roomsOption: [{id:1,value:1},{id:2,value:2},{id:3,value:3},{id:4,value:4},{id:5,value:5}],
        roomsCurrentValue:'',

        metroOption:[{id:1,value:'Грушевка'},{id:2,value:'Восход'},{id:3,value:'Пушкинская'},{id:4,value:'Партизанская'}],
        metroCurrentValue:'',

        districtOption:[{id:1,value:'Шабаны'},{id:2,value:'Запад'},{id:3,value:'Восток'},{id:4,value:'Малиновка'}],
        districtCurrentValue: '',

        productsLimit: 6,
        currentPage: 1,

        sortedOptions:[{id:1, value:'Сначала дешевые'}, {id:2, value: 'Сначала дорогие'}],
        sortedValue:'',


        bedSumOptions:[{id:1, value:'1'}, {id:2, value:'2'}, {id:3, value:'3'}, {id:4, value:'4'}, {id:5, value:'5'}],
        bedSumCurrentValue:'',

        checkboxsMoreOption:[{id:1, value:'Газовая плита', bol:false},  {id:2, value:'Духовка', bol:false}, {id:3, value:'Кофеварка', bol:false}, {id:4, value:'Микроволновая печь ', bol:false}, {id:5, value:'Посуда', bol:false}, {id:6, value:'Посудомоечная машина', bol:false}],

        minCost:'',
        maxCost:'',


        productsLen: 0,
        sortedProducts:[],
        products: [],
    },
    reducers: {
        fetchProducts(state, action){
            state.products = action.payload
        },
        setBedSum(state, action){
            state.bedSumCurrentValue = action.payload
        },

        checkboxHandler(state, action){
            state.checkboxsMoreOption.forEach(obj =>{
                if(obj.value.includes(action.payload)){
                    obj.bol = !obj.bol

                }
            })
        },

        likeHandler(state, action){
            state.products.forEach(obj=>{
                Object.keys(obj).forEach(key=>{
                    if(key === action.payload.type){
                        obj[key].forEach(e=>{
                            if (e.id === action.payload.id){
                                e.like = !e.like
                            }
                        })
                    }
                })
            })
        },
        setCity(state, action){
            state.cityCurrentValue = action.payload

        },
        setRooms(state, action){
            state.roomsCurrentValue = action.payload
        },
        setMinCost(state, action){
            if(action.payload<0 || action.payload > 1000){
                return
            }
            state.minCost = action.payload
        },
        setMaxCost(state, action){
            if(action.payload<0 || action.payload > 1000){
                return
            }
            state.maxCost = action.payload
        },
        setDistrict(state, action){
            state.districtCurrentValue = action.payload
        },
        setMetro(state, action){
            state.metroCurrentValue = action.payload
        },
        setCurrentPage(state, action){
            state.currentPage = action.payload
            window.scrollTo(0, 0)
        },
        setSortedValue(state, action){
            state.sortedValue = action.payload
        },

        setLimit(state, action){
            state.productsLimit = action.payload
        },
        resetFilter(state){
            state.sortedValue = ''
            state.roomsCurrentValue = ''
            state.minCost = ''
            state.maxCost = ''
            state.districtCurrentValue = ''
            state.metroCurrentValue = ''
            state.bedSumCurrentValue = ''
            state.checkboxsMoreOption.forEach(e=>{
                e.bol = false
            })
        },

        mainFilter(state, action){
            // Условия для высчитывания стоимости
            while(state.minCost > state.maxCost ){
                state.maxCost = +state.maxCost + 100
            }
            if(state.minCost === ''){
                state.minCost=0
            }
            if(state.maxCost === ''){
                state.maxCost = 1000
            }


            //Основная фильтрация
            let sortedRes= []
            state.products.forEach(obj =>{
                if (action.payload){
                    Object.keys(obj).forEach(key =>{
                        if(action.payload === key){
                            obj[key].forEach(e=>{
                                if(!state.roomsCurrentValue){
                                    if(+e.cost >= state.minCost && +e.cost <= state.maxCost && e.city.includes(state.cityCurrentValue) && e.district.includes(state.districtCurrentValue) && e.metro.includes(state.metroCurrentValue) && e.people.includes(state.bedSumCurrentValue)){
                                        sortedRes.push(e)
                                    }
                                } else if(+e.cost >= state.minCost && +e.cost <= state.maxCost && e.city.includes(state.cityCurrentValue) && e.rooms===state.roomsCurrentValue && e.district.includes(state.districtCurrentValue) && e.metro.includes(state.metroCurrentValue) && e.people.includes(state.bedSumCurrentValue)){
                                    sortedRes.push(e)
                                }
                            })
                        }
                    })
                } else{
                    Object.keys(obj).forEach(key =>{
                        obj[key].forEach(e=>{
                            if(!state.roomsCurrentValue){
                                if(+e.cost >= state.minCost && +e.cost <= state.maxCost && e.city.includes(state.cityCurrentValue) && e.district.includes(state.districtCurrentValue) && e.metro.includes(state.metroCurrentValue) && e.people.includes(state.bedSumCurrentValue)){
                                    sortedRes.push(e)
                                }
                            } else if(+e.cost >= state.minCost && +e.cost <= state.maxCost && e.city.includes(state.cityCurrentValue) && e.rooms===state.roomsCurrentValue && e.district.includes(state.districtCurrentValue) && e.metro.includes(state.metroCurrentValue) && e.people.includes(state.bedSumCurrentValue)){
                                sortedRes.push(e)
                            }
                        })
                    })
                }
            })
            state.productsLen = sortedRes.length

            // Сортировка массива
            switch (state.sortedValue) {
            case 'Сначала дешевые':
                sortedRes.sort((a, b)=>{
                    return a.cost - b.cost
                })
                break
            case 'Сначала дорогие':
                sortedRes.sort((a, b)=>{
                    return b.cost - a.cost
                })
                break
            default:
                break
            }

            // Фильтрация по страницам
            let res = []
            let test = 0
            for (let i = 0; test < sortedRes.length; i++) {
                let obj = []
                for (let j = 0; j< state.productsLimit; j++) {
                    if(sortedRes[test]){
                        obj.push(sortedRes[test])
                        test++
                    }

                }
                res.push(obj)
            }
            state.sortedProducts = res
            state.currentPage = 1
        },

    },
})

export default main.reducer
export const {setCity, setRooms, setMinCost, setMaxCost, mainFilter, setDistrict, setMetro, resetFilter, setCurrentPage, setSortedValue, setLimit, likeHandler, setBedSum, checkboxHandler, fetchProducts} = main.actions
