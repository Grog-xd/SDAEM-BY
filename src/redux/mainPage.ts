import {createSlice, PayloadAction} from '@reduxjs/toolkit';

import {ProductProps} from '../types/types';


interface likeActionProps{
    id: number,
    type:string
}

const main = createSlice({
    name: 'main',
    initialState:{
        tabs:[{id:'flats', value:'Квартиры'}, {id:'cottages', value:'Котеджы и усадьбы'}, {id:'baths', value:'Бани и сауны'},  {id:'cars', value:'Авто на прокат'}],

        cityOption:[{id:1,value:'Минск'},{id:2,value:'Гомель'},{id:3,value:'Брест'},{id:4,value:'Витебск'},{id:5,value:'Гродно'}, {id:6,value:'Могилев'}],
        cityCurrentValue:'',

        roomsOption: [{id:1,value:1},{id:2,value:2},{id:3,value:3},{id:4,value:4},{id:5,value:5}],
        roomsCurrentValue:'',

        metroOption:[{id:1,value:'Грушевка'},{id:2,value:'Восход'},{id:3,value:'Пушкинская'},{id:4,value:'Партизанская'}],
        metroCurrentValue:'',

        districtOption:[{id:1,value:'Шабаны'},{id:2,value:'Запад'},{id:3,value:'Восток'},{id:4,value:'Малиновка'}],
        districtCurrentValue: '',

        productsLimit: 6,
        currentPage: 1,

        sortedOptions:[{id:1, value:'Сначала дешевые'}, {id:2, value: 'Сначала дорогие'}],
        sortedValue:'',


        bedSumOptions:[{id:1, value:'1'}, {id:2, value:'2'}, {id:3, value:'3'}, {id:4, value:'4'}, {id:5, value:'5'}],
        bedSumCurrentValue:'',

        checkboxsMoreOption:[{id:1, value:'Газовая плита', bol:false},  {id:2, value:'Духовка', bol:false}, {id:3, value:'Кофеварка', bol:false}, {id:4, value:'Микроволновая печь ', bol:false}, {id:5, value:'Посуда', bol:false}, {id:6, value:'Посудомоечная машина', bol:false}],

        minCost:0,
        maxCost:1000,


        productsLen: 0,
        sortedProducts:[],
        products: [],
    },
    reducers: {
        fetchProducts(state, action){
            state.products = action.payload
        },
        setBedSum(state, action:PayloadAction<string>){
            state.bedSumCurrentValue = action.payload
        },

        checkboxHandler(state, action:PayloadAction<string>){
            state.checkboxsMoreOption.forEach(obj =>{
                if(obj.value.includes(action.payload)){
                    obj.bol = !obj.bol

                }
            })
        },

        likeHandler(state, action:PayloadAction<likeActionProps>){
            state.products.forEach(obj=>{
                Object.keys(obj).forEach(key=>{
                    if(key === action.payload.type){
                        obj[key].forEach(e=>{
                            if (e.id === action.payload.id){
                                e.like = !e.like
                            }
                        })
                    }
                })
            })
        },
        setCity(state, action:PayloadAction<string>){
            state.cityCurrentValue = action.payload

        },
        setRooms(state, action:PayloadAction<string>){
            state.roomsCurrentValue = action.payload
        },
        setMinCost(state, action:PayloadAction<number>){
            if(action.payload<0 || action.payload > 1000){
                return
            }
            state.minCost = action.payload
        },
        setMaxCost(state, action:PayloadAction<number>){
            if(action.payload<0 || action.payload > 1000){
                return
            }
            state.maxCost = action.payload
        },
        setDistrict(state, action:PayloadAction<string>){
            state.districtCurrentValue = action.payload
        },
        setMetro(state, action:PayloadAction<string>){
            state.metroCurrentValue = action.payload
        },
        setCurrentPage(state, action:PayloadAction<number>){
            state.currentPage = action.payload
            window.scrollTo(0, 0)
        },
        setSortedValue(state, action:PayloadAction<string>){
            state.sortedValue = action.payload
        },

        setLimit(state, action:PayloadAction<number>){
            state.productsLimit = action.payload
        },
        resetFilter(state){
            state.sortedValue = ''
            state.roomsCurrentValue = ''
            state.minCost = undefined
            state.maxCost = undefined
            state.districtCurrentValue = ''
            state.metroCurrentValue = ''
            state.bedSumCurrentValue = ''
            state.checkboxsMoreOption.forEach(e=>{
                e.bol = false
            })
        },

        mainFilter(state, action:PayloadAction<ProductProps>){
            // Условия для высчитывания стоимости
            while(state.minCost > state.maxCost ){
                state.maxCost = state.maxCost + 100
            }
            if(state.minCost === undefined){
                state.minCost = 0
            }
            if(state.maxCost === undefined){
                state.maxCost = 1000
            }


            //Основная фильтрация
            let sortedRes= []
            state.products.forEach(obj =>{
                if (action.payload){
                    Object.keys(obj).forEach(key =>{
                        if(action.payload === key){
                            obj[key].forEach(e=>{
                                if(!state.roomsCurrentValue){
                                    if(+e.cost >= state.minCost && +e.cost <= state.maxCost && e.city.includes(state.cityCurrentValue) && e.district.includes(state.districtCurrentValue) && e.metro.includes(state.metroCurrentValue) && e.people.includes(state.bedSumCurrentValue)){
                                        sortedRes.push(e)
                                    }
                                } else if(+e.cost >= state.minCost && +e.cost <= state.maxCost && e.city.includes(state.cityCurrentValue) && e.rooms===state.roomsCurrentValue && e.district.includes(state.districtCurrentValue) && e.metro.includes(state.metroCurrentValue) && e.people.includes(state.bedSumCurrentValue)){
                                    sortedRes.push(e)
                                }
                            })
                        }
                    })
                } else{
                    Object.keys(obj).forEach(key =>{
                        obj[key].forEach(e=>{
                            if(!state.roomsCurrentValue){
                                if(+e.cost >= state.minCost && +e.cost <= state.maxCost && e.city.includes(state.cityCurrentValue) && e.district.includes(state.districtCurrentValue) && e.metro.includes(state.metroCurrentValue) && e.people.includes(state.bedSumCurrentValue)){
                                    sortedRes.push(e)
                                }
                            } else if(+e.cost >= state.minCost && +e.cost <= state.maxCost && e.city.includes(state.cityCurrentValue) && e.rooms===state.roomsCurrentValue && e.district.includes(state.districtCurrentValue) && e.metro.includes(state.metroCurrentValue) && e.people.includes(state.bedSumCurrentValue)){
                                sortedRes.push(e)
                            }
                        })
                    })
                }
            })
            state.productsLen = sortedRes?.length

            // Сортировка массива
            switch (state.sortedValue) {
            case 'Сначала дешевые':
                sortedRes.sort((a, b)=>{
                    return a.cost - b.cost
                })
                break
            case 'Сначала дорогие':
                sortedRes.sort((a, b)=>{
                    return b.cost - a.cost
                })
                break
            default:
                break
            }

            // Фильтрация по страницам
            let res = []
            let test = 0
            for (let i = 0; test < sortedRes?.length; i++) {
                let obj = []
                for (let j = 0; j< state.productsLimit; j++) {
                    if(sortedRes[test]){
                        obj.push(sortedRes[test])
                        test++
                    }

                }
                res.push(obj)
            }
            state.sortedProducts = res
            state.currentPage = 1
        },

    },
})

export default main.reducer
export const {setCity, setRooms, setMinCost, setMaxCost, mainFilter, setDistrict, setMetro, resetFilter, setCurrentPage, setSortedValue, setLimit, likeHandler, setBedSum, checkboxHandler, fetchProducts} = main.actions
