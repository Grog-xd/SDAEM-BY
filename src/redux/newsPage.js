import {createSlice} from "@reduxjs/toolkit";

const newsPage = createSlice({
    name: 'news',
    initialState:{
        sortedPosts:[],
        limitNewsItems:9,
        currentPage:1,
        contactsInformation:{location:'220068, РБ, г. Минск, ул. Осипенко, 21, кв.23', tel:'+375 29 621-48-33', email:'sdaem@sdaem.by', timeWork:'08:00-22:00'},
        posts:[],
    },
    reducers:{
        fetchNews(state, action){
            state.posts = action.payload
        },
        getNews(state, action){
            let sortedPosts = state.posts.filter(post=> post.title.toLowerCase().includes(action.payload.toLowerCase()))

            let res = []
            let test = 0
            while(test < sortedPosts.length) {
                let obj = []
                while(obj.length < state.limitNewsItems){

                    if(test >= sortedPosts.length){
                        break
                    }
                    obj.push(sortedPosts[test])
                    test++
                }
                res.push(obj)
            }
            state.sortedPosts = res
            state.currentPage = 1
        },
        setNewsPage(state, action){
            state.currentPage = action.payload
            window.scrollTo(0, 0)
        },

    }
})

export default newsPage.reducer
export const {setNewsPage, getNews, fetchNews} = newsPage.actions
