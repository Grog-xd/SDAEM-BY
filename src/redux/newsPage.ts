import {createSlice, PayloadAction} from '@reduxjs/toolkit';

import {newProps} from '../types/types';

const newsPage = createSlice({
    name: 'news',
    initialState:{
        sortedPosts:[],
        limitNewsItems:9,
        currentPage:1,
        contactsInformation:{location:'220068, РБ, г. Минск, ул. Осипенко, 21, кв.23', tel:'+375 29 621-48-33', email:'sdaem@sdaem.by', timeWork:'08:00-22:00'},
        posts:[],
    },
    reducers:{
        fetchNews(state, action:PayloadAction<newProps[]>){
            state.posts = action.payload
        },
        getNews(state, action:PayloadAction<string>){
            let sortedPosts = state.posts.filter(post=> post.title.toLowerCase().includes(action.payload.toLowerCase()))

            let res = []
            let test = 0
            while(test < sortedPosts.length) {
                let obj = []
                while(obj.length < state.limitNewsItems){

                    if(test >= sortedPosts.length){
                        break
                    }
                    obj.push(sortedPosts[test])
                    test++
                }
                res.push(obj)
            }
            state.sortedPosts = res
            state.currentPage = 1
        },
        setNewsPage(state, action:PayloadAction<number>){
            state.currentPage = action.payload
            window.scrollTo(0, 0)
        },

    },
})

export default newsPage.reducer
export const {setNewsPage, getNews, fetchNews} = newsPage.actions
