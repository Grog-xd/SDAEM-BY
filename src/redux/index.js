import {combineReducers} from "redux";
import {configureStore} from "@reduxjs/toolkit";
import mainPage from "./mainPage";
import newsPage from "./newsPage";
import loginPage from "./loginPage";

const rootReducer = combineReducers({
    news: newsPage,
    main:mainPage,
    login: loginPage
})

export const store = configureStore({
    reducer: rootReducer
})
