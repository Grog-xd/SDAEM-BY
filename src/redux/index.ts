import {combineReducers} from 'redux';
import {configureStore} from '@reduxjs/toolkit';
import {useDispatch} from 'react-redux';

import mainPage from './mainPage';
import newsPage from './newsPage';
import loginPage from './loginPage';

const rootReducer = combineReducers({
    news: newsPage,
    main:mainPage,
    login: loginPage,
})

export const store = configureStore({
    reducer: rootReducer,
})


export type AppDispatch = typeof store.dispatch
export const useAppDispatch: () => AppDispatch = useDispatch
