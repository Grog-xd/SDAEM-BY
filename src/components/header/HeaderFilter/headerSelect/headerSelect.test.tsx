import {render, screen} from '@testing-library/react';
import {Provider} from 'react-redux';

import userEvent from '@testing-library/user-event';

import {MemoryRouter} from 'react-router-dom';

import {store} from '../../../../redux';

import HeaderSelect from './headerSelect';

describe('header select tests', () =>{
    test('snapshot test', ()=>{
        render(
            <Provider store={store} >
                <MemoryRouter>
                    <HeaderSelect type={'flats'} value={'Квартиры на сутки'}  />
                </MemoryRouter>
            </Provider>,
        )
        const select = screen.getByTestId('header-select')
        expect(select).toMatchSnapshot();
    })

    test('open options test', ()=>{
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <HeaderSelect type={'flats'} value={'Квартиры на сутки'}  />
                </MemoryRouter>
            </Provider>,
        )
        const headerSelectBtn = screen.getByTestId('header-select-btn')

        expect(screen.queryByTestId('header-options-block')).not.toBeInTheDocument()
        userEvent.click(headerSelectBtn)
        expect(screen.getByTestId('header-options-block')).toBeInTheDocument()
    })
})
