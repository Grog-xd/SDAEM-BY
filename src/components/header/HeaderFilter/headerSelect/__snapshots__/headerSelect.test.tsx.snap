// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header select tests snapshot test 1`] = `
<div
  class="select"
  data-testid="header-select"
>
  <button
    class="changeBlock"
    data-testid="header-select-btn"
  >
    Квартиры на сутки
  </button>
</div>
`;
